javalib/src/main/scala/java/net/URI.scala

```diff
@@ -107,16 +107,22 @@ final class URI(origStr: String) extends Serializable with Comparable[URI] {
       else x.fold(-1)(s1 => y.fold(1)(s2 => cmp(s1, s2)))
     }
 
-    if (this._scheme != that._scheme)
-      this._scheme.fold(-1)(s1 => that._scheme.fold(1)(s1.compareToIgnoreCase))
-    else if (this._isOpaque != that._isOpaque)
-      // A hierarchical URI is less than an opaque URI
-      if (this._isOpaque) 1 else -1
-    else if (_isOpaque) {
+    @inline def cmpSsp(): Int =  {
       val ssp = cmp(this._schemeSpecificPart, that._schemeSpecificPart)
       if (ssp != 0) ssp
       else cmpOpt(this._fragment, that._fragment)
-    } else if (this._authority != that._authority) {
+    }
+
+    if (this._scheme != that._scheme) {
+      val s = this._scheme.fold(-1)(s1 => that._scheme.fold(1)(s1.compareToIgnoreCase))
+      if (s == 0 && isOpaque()) cmpSsp()
+      else s
+    } else if (this._isOpaque != that._isOpaque)
+      // A hierarchical URI is less than an opaque URI
+      if (this._isOpaque) 1 else -1
+    else if (_isOpaque)
+      cmpSsp()
+    else if (this._authority != that._authority) {
```

Comment:
While you are here, please put braces around all if / else blocks (unfortunately this seems undocumented but is a style we follow).

Reply:
Fixed